sdk/samples/core/rpc_client_pending_commands.h

```diff
@@ -0,0 +1,151 @@
+// Copyright (c) Microsoft Corporation. All rights reserved.
+// SPDX-License-Identifier: MIT
+
+/**
+ * @file
+ *
+ * @brief Definition of #az_mqtt5_rpc_client_sample pending command functions. You use the RPC
+ * client to send commands.
+ *
+ * @note You MUST NOT use any symbols (macros, functions, structures, enums, etc.)
+ * prefixed with an underscore ('_') directly in your application code. These symbols
+ * are part of Azure SDK's internal implementation; we do not document these symbols
+ * and they are subject to change in future versions of the SDK which would break your code.
+ */
+
+#ifndef _RPC_CLIENT_PENDING_COMMANDS_H
+#define _RPC_CLIENT_PENDING_COMMANDS_H
+
+#include <azure/core/az_mqtt5_connection.h>
+#include <azure/core/az_result.h>
+#include <azure/core/az_span.h>
+
+#include <azure/core/_az_cfg_prefix.h>
+
+#ifndef RPC_CLIENT_MAX_PENDING_COMMANDS
+#define RPC_CLIENT_MAX_PENDING_COMMANDS 5
+#endif
+
+typedef struct pending_command
+{
+  az_span correlation_id;
+  az_context context;
+  az_span command_name;
+} pending_command;
+
+typedef struct pending_commands_array
+{
+  int32_t pending_commands_count;
+
+  pending_command commands[RPC_CLIENT_MAX_PENDING_COMMANDS];
+} pending_commands_array;
+
+AZ_INLINE az_result pending_commands_array_init(
+    pending_commands_array* pending_commands,
+    uint8_t correlation_id_buffers[RPC_CLIENT_MAX_PENDING_COMMANDS]
+                                  [AZ_MQTT5_RPC_CORRELATION_ID_LENGTH])
+{
+  pending_commands->pending_commands_count = 0;
+  for (int i = 0; i < RPC_CLIENT_MAX_PENDING_COMMANDS; i++)
+  {
+    pending_commands->commands[i].correlation_id
+        = az_span_create(correlation_id_buffers[i], AZ_MQTT5_RPC_CORRELATION_ID_LENGTH);
+    az_span_fill(pending_commands->commands[i].correlation_id, 0x0);
+    pending_commands->commands[i].command_name = AZ_SPAN_EMPTY;
+  }
+  return AZ_OK;
+}
+
+AZ_INLINE az_result add_command(
+    pending_commands_array* pending_commands,
+    az_span correlation_id,
+    az_span command_name,
+    int32_t timout_ms)
+{
+  az_result ret = AZ_OK;
+  if (pending_commands->pending_commands_count >= RPC_CLIENT_MAX_PENDING_COMMANDS)
+  {
+    printf(LOG_APP_ERROR "Pending Commands Array already has max number of commands.\n");
+    return AZ_ERROR_OUT_OF_MEMORY;
+  }
+  int32_t empty_index = -1;
+  for (int i = 0; i < RPC_CLIENT_MAX_PENDING_COMMANDS; i++)
+  {
+    if (az_span_size(pending_commands->commands[i].command_name) == 0)
+    {
+      empty_index = i;
+      break;
+    }
+  }
+  if (empty_index < 0)
```

Comment:
This is mostly for safety - if pending_commands_count is less than max, then there _should_ be an available spot in the array, but just in case it couldn't find one, I want to protect against indexing at -1

Reply:
Use assert instead?

Reply:
I don't see an assert function that we use elsewhere in the code outside of tests - it seems most places we use preconditions or if statements - let me know if there is something we usually use!

Reply:
can this be a precondition?

Reply:
Doesn't seem to be a precondition candidate, since the array can change during the execution of the program.

Reply:
will leave this as is